lua_modules/sntp/sntp.lua

```diff
@@ -0,0 +1,290 @@
+local MAX_SERVER_ATTEMPTS = 2
+local SNTP_TIMEOUT        = 5000
+
+-- Constructor
+--   sc and fc are our Success and Failure Callbacks, resp.
+local new = function(serv, sc, fc, now)
+
+  if     type(serv) == "string" then serv = {serv}
+  elseif serv == nil            then serv =
+    {
+      nil,
+      "1.nodemcu.pool.ntp.org",
+      "2.nodemcu.pool.ntp.org",
+    }
+    local ni = net.ifinfo(0)
+    ni = ni and ni.dhcp
+    serv[1] = ni.ntp_server or "0.nodemcu.pool.ntp.org"
+  elseif type(serv) ~= "table"  then error "Bad server table"
+  end
+
+  if type(sc) ~= "function" and type(sc) ~= "lightfunction" then
+    error "Bad success callback type"
+  end
+  if fc ~= nil and type(fc) ~= "function" and type(fc) ~= "lightfunction" then
+    error "Bad failure callback type"
+  end
+  if now ~= nil and type(now) ~= "function" and type(now) ~= "lightfunction" then
+    error "Bad clock type"
+  end
+  now = now or (rtctime and rtctime.get)
+  if now == nil then error "Need clock function" end
+
+  local self = {}
```

Comment:
Incidentally why not `local self = {servers = serv}` ?

I more general point given that self is a module local, this is a singleton class  when I am not sure this is the intent.  Shouldn't self be local to `new()` so 
```Lua
local sntp = require'sntp'.open(a,b,c,d) works as expected
```
Also in general the object should be the first param so the ":" form of method invocation works as expected.

Reply:
`self` *is* local to `new`?  The only module locals are `MAX_SERVER_ATTEMPTS`, `SNTP_TIMEOUT`, `new`, `update_rtc`, and `go`.  Or am I missing something?

While I generally understand the `:` calling convention, I don't see its relevance here?  Neither of the exported functions on the "SNTP object" created by `new` (i.e., `sync()` and `stop()`) take arguments, as all their associated state is in their upvals.  The callbacks are held as upvals not as table members.

Reply:
Sorry, yes self is as you say, but I got a bit lost in the depth of semantic nesting.  For others, a call `v:name(args)` is syntactic sugar for `v.name(v,args)`, except that this takes less VM instructions and v is evaluated only once.

Here is the size / upvalue analysis of the module (hope you find this useful).
```
main     <L0,0>     (15 ins, 5 const)     
function <L6,201>   (145 ins, 25 const) -- upvals:  SNTP_TIMEOUT, MAX_SERVER_ATTEMPTS
function <L45,63>   (50 ins, 9 const)   -- upvals:  _tmr, _six, _sat, _res, _best, _udp, _kod
function <L71,143>  (42 ins, 10 const)  -- upvals: _tmr, SNTP_TIMEOUT, _udp, doserver, now, _kod, fc, serv, _six, _self, nextServer,
                                                    _pbest, _res, _best
function <L72,76>   (15 ins, 4 const)   -- upvals: _udp, doserver
function <L82,137>  (142 ins, 21 const) -- upvals: ip, txts, now, _kod, fc, serv, _six, _self, nextServer, _pbest, _res, _best, _tmr
function <L146,157> (31 ins, 2 const)   -- upvals: _sat, MAX_SERVER_ATTEMPTS, fc, serv, _six, _self, nextServer, _udp, doserver, hail
function <L153,156> (16 ins, 3 const)   -- upvals: doserver, hail
function <L162,179> (45 ins, 4 const)   -- upvals: _six, serv, _res, _pbest, _best, _stop, sc, _self, fc, _sat, doserver
function <L184,191> (18 ins, 6 const)   -- upvals: _stop, _udp, _tmr, _six, nextServer
function <L193,197> (11 ins, 1 const)   -- upvals: _res, _best, _stop
function <L204,246> (58 ins, 16 const)             
function <L251,273> (26 ins, 6 const)   -- upvals: new, update_rtc
function <L252,252> (1 ins, 0 const) 
function <L254,266> (40 ins, 9 const)   -- upvals: update_rtc, freq, sc
function <L270,270> (6 ins, 2 const)    -- upvals: sntpobj
```
This is a pretty chunky piece of code that takes 10kB RAM to load if you are using LFS.  That's a big enough chunk to think about how we could optimise this.  Have a look at how I've done this for the FTP server.

Reply:
I assume that's supposed to be "10kB RAM if you're **not** using LFS"?  If this is 10K even with LFS then I really am in trouble.  I'll see what I can do to drive it down all the same.  (I continue to vaguely wish that Lua had a `const` keyword.)

Again, I know what `:` does.  I know it's syntactic sugar and I understand the VM instructions it expands into.  What I don't understand is why you bring it up.  Am I `foo.bar(foo,...)`-ing somewhere in here?  Am I failing to provide an interface that looks like it should support that?

Reply:
> I assume that's supposed to be "10kB RAM if you're **not** using LFS"?

Correct.  The static RAM usage if loaded into LFS is zero, though executing code does create runtime variables.  Lua constants are just that and are stored as TValues in the code space (and which is why I listed them above).   Each instruction takes 4 bytes, each TValue takes 12 bytes in Lua51 and 8 bytes in Lua53.  `Proto` structures and (debug) `lineinfo` vectors also take code space, though the NodeMCU lineinfo is typically 10-20× more dense than in standard Lua).  String constants are a TValue pointing to a ROstrt entry.  

Unlike PHP or Javascript, Lua has no concept of an array constant; arrays are constructed in RAM at runtime, though the opcodes which do this construction are instruction that can be in LFS.  The only exception to this was introduced by eLua and built upon in NodeMCU; this is the ROTable construct, but these are declared statically in C code as `const` structures and mapped into Flash memory.